const-generic-const-fn-bounds.md

```diff
@@ -0,0 +1,496 @@
+}
+```
+
+Where all trait impls *must* provide a `const` function for `a`, allowing
+
+```rust
+const fn foo<T: ?const Foo>() -> i32 {
+    T::a()
+}
+```
+
+even though the `?const` modifier explicitly opts out of constness.
+
+The author of this RFC believes this feature to be unnecessary, since one can get the same effect
+by splitting the trait into its const and nonconst parts:
+
+```rust
+trait FooA {
+    fn a() -> i32;
+}
+trait FooB {
+    fn b() -> i32;
+}
+const fn foo<T: FooA + ?const FooB>() -> i32 {
+    T::a()
+}
+```
+
+Impls of the two traits can then decide constness of either impl at their leasure.
+
+### `const` traits
+
+A further extension could be `const trait` declarations, which desugar to all methods being `const`:
+
+```rust
+const trait V {
+    fn foo(C) -> D;
+    fn bar(E) -> F;
+}
+// ...desugars to...
+trait V {
+    const fn foo(C) -> D;
+    const fn bar(E) -> F;
+}
+```
+
+## `?const` modifiers in trait methods
+
+This RFC does not touch `trait` methods at all, all traits are defined as they would be defined
+without `const` functions existing. A future extension could allow
+
+```rust
+trait Foo {
+    fn a<T: ?const Bar>() -> i32;
+}
+```
+
+which does not force `impl const Foo for Type` to now require passing a `T` with an `impl const Bar`
+to the `a` method.
+
+## `const` function pointers
+
+```rust
+const fn foo(f: fn() -> i32) -> i32 {
+    f()
+}
+```
+
+is currently illegal. While we can change the language to allow this feature, two questions make
+themselves known:
+
+1. fn pointers in constants
+
+    ```rust
+    const F: fn() -> i32 = ...;
+    ```
+
+    is already legal in Rust today, even though the `F` doesn't need to be a `const` function.
+
+2. Opt out bounds are ugly
+
+    I don't think it's either intuitive nor readable to write the following
+
+    ```rust
+    const fn foo(f: ?const fn() -> i32) -> i32 {
+        // not allowed to call `f` here, because we can't guarantee that it points to a `const fn`
+    }
+    ```
+
+Thus it seems useful to prefix function pointers to `const` functions with `const`:
+
+```rust
+const fn foo(f: const fn() -> i32) -> i32 {
+    f()
+}
+const fn bar(f: fn() -> i32) -> i32 {
+    f() // ERROR
+}
+```
+
+This opens up the curious situation of `const` function pointers in non-const functions:
+
+```rust
+fn foo(f: const fn() -> i32) -> i32 {
+    f()
+}
+```
+
+Which is useless except for ensuring some sense of "purity" of the function pointer ensuring that
+subsequent calls will only modify global state if passed in via arguments.
+
+## explicit `const` bounds
+
+`const` on the bounds (e.g. `T: const Trait`) requires an `impl const Trait` for any types used to
+replace `T`. This allows `const` trait bounds on any (even non-const) functions, e.g. in
+
+```rust
+fn foo<T: const Bar>() -> i32 {
+    const FOO: i32 = T::bar();
+    FOO
+}
+```
+
+Which, once `const` items and array lengths inside of functions can make use of the generics of
```

Comment:
Is there a reason generics cannot be used inside `const` contexts?

Also, even without directly accessing the parameters, we have things like:
```rust
fn foo<T: const Bar>(t: T) -> i32 {
    const FOO: i32 = t.baz();
    FOO
}
```
This seems like an addition that goes hand-in-hand with the rest of the RFC.

Reply:
> Also, even without directly accessing the parameters, we have things like:

That is impossible, the `t` itself is a runtime value, you can't have a compile-time value depend on a runtime value. `const` items aren't the same thing as immutable let bindings.

> Is there a reason generics cannot be used inside const contexts?

Basically the reason is that we don't have a concept of generic modules, and items declared inside a function body are essentially inside their own anonymous module. It's entirely unrelated to const contexts and all about items. You can't do

```rust
fn foo<T>(t: T) -> i32 {
    fn bar(t: T) -> i32 {
        42
    }
    bar()
}
```

Array lengths are an anonymous const item and thus subject to the same restrictions that other items are

Reply:
> That is impossible, the t itself is a runtime value, you can't have a compile-time value depend on a runtime value.

Oh, d'uh, of course.

> Basically the reason is that we don't have a concept of generic modules, and items declared inside a function body are essentially inside their own anonymous module.

Okay, so it's an implementation detail more than anything else. In that case, it's not necessary to have `const` bounds until that's possible.